### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by ghost**
This template has become a bit outdated for my usage, so I decided to rewrite it. While the new template works, it needs better testing, please do not merge this right away if you decide to.

Features added:

- Full support for Quilt
- Example mixins (all platforms), example common mixin applied to all of them
- Access wideners for Fabric and Quilt (disabled by default)

Changes:

- Updated Gradle to 8.1.1
- Updated plugins and dependencies
- Mod domain can now be changed from properties
- Ability to change Java version (technical, for older MC versions)

Fixes:

- Some boolesn properties not actually working (were always acting as `true`)
- Common sources and Javadoc were not added to mod-sources and mod-javadoc jars respectively

NYI/To do:

- CI
- Tweak Maven publishing (script needs updating)
- Figure out what to do with different mappings support

### CODE REVIEW

File: Common/build.gradle [hunk 1]
```diff
@@ -1,37 +1,37 @@
...
+            client(common_client_run_name as String) {
                 workingDirectory(this.file("run"))
             }
         }
     }
 }
 
 dependencies {
-     compileOnly group:'org.spongepowered', name:'mixin', version:'0.8.5'
-     implementation group: 'com.google.code.findbugs', name: 'jsr305', version: '3.0.1'
```
Comment (jaredlll08) [inline Common/build.gradle:26]: Removing JSR305 is going to be a headache for authors when their IDE inevitably adds the import automatically.

File: Common/build.gradle [hunk 2]
```diff
@@ -1,37 +1,37 @@
...
+    compileOnly "org.spongepowered:mixin:${mixin_version}"
 }
 
-processResources {
+// Add the generated resources to the main source set
+sourceSets.main.resources { srcDir 'src/generated/resources' }
 
-    def buildProps = project.properties.clone()
+processResources {
+    def buildProps = new HashMap<>(project.properties)
```
Comment (jaredlll08) [inline Common/build.gradle:33]: How is this better than `project.properties.clone()`?

File: Common/src/main/java/com/example/examplemod/CommonInit.java [hunk 3]
```diff
@@ -8,24 +8,24 @@
 // import and access the vanilla codebase, libraries used by vanilla, and optionally third party libraries that provide
 // common compatible binaries. This means common code can not directly use loader specific concepts such as Forge events
 // however it will be compatible with all supported mod loaders.
-public class CommonClass {
+public class CommonInit {
```
Comment (jaredlll08) [inline Common/src/main/java/com/example/examplemod/CommonInit.java:11]: Why the name change?

File: Common/src/main/java/com/example/examplemod/CommonInit.java [hunk 4]
```diff
@@ -8,24 +8,24 @@
...
         // It is common for all supported loaders to provide a similar feature that can not be used directly in the
         // common code. A popular way to get around this is using Java's built-in service loader feature to create
         // your own abstraction layer. You can learn more about this in our provided services class. In this example
         // we have an interface in the common code and use a loader specific implementation to delegate our call to
         // the platform specific approach.
-        if (Services.PLATFORM.isModLoaded("examplemod")) {
+        if (Services.PLATFORM.isModLoaded(Constants.MOD_ID)) {
 
-            Constants.LOG.info("Hello to examplemod");
+            Constants.LOGGER.info("Hello to " + Constants.MOD_ID);
```
Comment (jaredlll08) [inline Common/src/main/java/com/example/examplemod/CommonInit.java]: This is not the intended function of this code.

The code is meant to show how you can check if *another* mod is loaded and say hello to that mod, not saying hello to ourselves.

File: Common/src/main/java/com/example/examplemod/Constants.java [hunk 5]
```diff
@@ -5,7 +5,11 @@
...
-	public static final String MOD_ID = "multiloader";
-	public static final String MOD_NAME = "Multi Loader Template";
-	public static final Logger LOG = LoggerFactory.getLogger(MOD_NAME);
+	public static final String MOD_ID = "examplemod";
+	public static final String MOD_NAME = "Example Mod";
+
+	// This logger is used to write text to the console and the log file.
+	// It is considered best practice to use your mod id as the logger's name.
+	// That way, it's clear which mod wrote info, warnings, and errors.
+	public static final Logger LOGGER = LoggerFactory.getLogger(MOD_ID);
```
Comment (jaredlll08) [inline Common/src/main/java/com/example/examplemod/Constants.java:14]: Why the name change?

File: Common/src/main/resources/pack.mcmeta
Comment (jaredlll08) [inline Common/src/main/resources/pack.mcmeta:1]: You've removed this file but haven't added it to fabric or quilt?

File: Fabric/build.gradle [hunk 6]
```diff
@@ -1,21 +1,25 @@
 plugins {
-    id 'fabric-loom' version '1.0-SNAPSHOT'
-    id 'maven-publish'
+    id 'java'
     id 'idea'
+	id 'fabric-loom' version '1.1.+'
```
Comment (jaredlll08) [inline Fabric/build.gradle]: Why do you keep the version specification here but not for the other plugins?

File: Fabric/build.gradle [hunk 7]
```diff
@@ -29,21 +33,98 @@ loom {
...
+        from project(":Common").sourceSets.main.allSource
     }
 }
 
-tasks.withType(JavaCompile) {
-    source(project(":Common").sourceSets.main.allSource)
+javadoc {
+    if (Boolean.parseBoolean(common_add_to_javadoc)) {
+        source(project(":Common").sourceSets.main.allJava)
+    }
```
Comment (jaredlll08) [inline Fabric/build.gradle:119]: Why is this an option??

File: Fabric/src/main/java/com/example/examplemod/mixin/ExampleMixin.java
Comment (jaredlll08) [inline Fabric/src/main/java/com/example/examplemod/mixin/ExampleMixin.java:1]: Why the change in mixin target?

File: Fabric/src/main/resources/examplemod.mixins.json
Comment (jaredlll08) [inline Fabric/src/main/resources/examplemod.mixins.json:1]: you named the other mixin file `examplemod-common.mixins.json` but this one just `examplemod.mixins.json`?

If anything it should be the other way around, common mixins is just `examplemod.mixins.json` and this `examplemod.fabric.mixins.json`

File: Forge/build.gradle [hunk 8]
```diff
@@ -1,27 +1,51 @@
...
+        case "official" :
+            mappings channel: 'official', version: minecraft_version
+            break
+        case "parchment" :
+            mappings channel: 'parchment', version: "${parchment_mc_version}-${parchment_mappings}-${minecraft_version}"
+            break
+        default :
+            mappings channel: 'official', version: minecraft_version
+            break
+    }
```
Comment (jaredlll08) [inline Forge/build.gradle]: the `default` and `official` case can be collapsed.

File: Forge/build.gradle [hunk 9]
```diff
@@ -1,27 +1,51 @@
...
+        println('Forge Access Transformers are enabled for this project.\n')
     }
 
     runs {
         client {
             workingDirectory project.file('run')
             ideaModule "${rootProject.name}.${project.name}.main"
             taskName 'Client'
+            property 'forge.logging.markers', 'REGISTRIES'
+            property 'forge.logging.console.level', 'debug'
```
Comment (jaredlll08) [inline Forge/build.gradle]: These markers were specifically removed.

Modders don't read their own log files unless they are looking for a specific message that they print out.

You may not agree with that statement, but I have dealt with dozens of issues about JEI not reloading and in the log file I see countless errors from JEI telling mods their recipes or plugins are broken, the mods JEI complains about are not small mods either, they are some of the biggest mods.

Having these lines just causes more log spam for you to look through to find the printout that you are actually looking for.

File: Forge/build.gradle [hunk 10]
```diff
@@ -63,23 +109,43 @@ minecraft {
     }
 }
 
-sourceSets.main.resources.srcDir 'src/generated/resources'
+//createMcpToSrg.finalizedBy('genIntellijRuns')
```
Comment (jaredlll08) [inline Forge/build.gradle:120]: What is this line meant to achieve (if uncommented)

File: Forge/build.gradle [hunk 11]
```diff
@@ -63,23 +109,43 @@ minecraft {
...
 dependencies {
     minecraft "net.minecraftforge:forge:${minecraft_version}-${forge_version}"
-    compileOnly project(":Common")
-}
 
-tasks.withType(JavaCompile) {
-    source(project(":Common").sourceSets.main.allSource)
+    annotationProcessor "org.spongepowered:mixin:${mixin_version}:processor"
+
+    implementation project(":Common")
```
Comment (jaredlll08) [inline Forge/build.gradle:130]: I forgot the reason, but there was a very specific reason I went with `compileOnly` instead of `implementation`.

I think forge pulls the code into runtime through the runconfigs, so having this as implementation means that there is duplicate classes.

File: Forge/build.gradle [hunk 12]
```diff
@@ -63,23 +109,43 @@ minecraft {
...
+    if (Boolean.parseBoolean(common_add_to_sources)) {
+        from project(":Common").sourceSets.main.allSource
+    }
+}
+
+javadoc {
+    if (Boolean.parseBoolean(common_add_to_javadoc)) {
+        source(project(":Common").sourceSets.main.allJava)
+    }
+}
```
Comment (jaredlll08) [inline Forge/build.gradle:151]: Why is this an option??

File: Forge/src/main/java/com/example/examplemod/ExampleMod.java
Comment (jaredlll08) [inline Forge/src/main/java/com/example/examplemod/ExampleMod.java:1]: While this may be the contents of forge's MDK, it doesn't make sense for this project, as you need more boilerplate to register blocks and items.

File: Forge/src/main/resources/examplemod.mixins.json
Comment (jaredlll08) [inline Forge/src/main/resources/examplemod.mixins.json:1]: The same comment about fabric's mixins.json applies here.

File: build.gradle [hunk 13]
```diff
@@ -1,63 +1,87 @@
+plugins {
+    id 'java'
+    id 'org.spongepowered.gradle.vanilla' version '0.2.+' apply false
+    id 'net.minecraftforge.gradle' version '5.1.+' apply false
+    id 'org.spongepowered.mixin' version '0.7.+' apply false
+    id 'org.parchmentmc.librarian.forgegradle' version '1.+' apply false
+    //id 'fabric-loom' version '1.1.+' apply false
+    //id 'org.quiltmc.loom' version '1.1+' apply false
```
Comment (jaredlll08) [inline build.gradle]: Why commented out?

File: build.gradle [hunk 14]
```diff
@@ -1,63 +1,87 @@
...
-    java.withJavadocJar()
-        
+
+    group "${mod_domain}.${mod_author}.${mod_id}"
+    version mod_version
+
+    java {
+        toolchain.languageVersion.set(JavaLanguageVersion.of(java_version as int))
+        //setSourceCompatibility(JavaVersion.toVersion(java_version as int))
+        //setTargetCompatibility(JavaVersion.toVersion(java_version as int))
```
Comment (jaredlll08) [inline build.gradle]: as far as I'm aware, these lines have been replaced by the line above.

File: gradle.properties [hunk 15]
```diff
@@ -1,26 +1,54 @@
-# Project
-version=1.0.0
-group=com.example.examplemod
+# Mod
+mod_id=examplemod
+mod_name=Example Mod
+mod_version=1.0.0
+mod_author=example
+mod_domain=com
```
Comment (jaredlll08) [inline gradle.properties]: This is not a good idea. 

Not everyone uses the domain convention for packages, for example `JEI` uses `mezz.jei`.

File: gradle.properties [hunk 16]
```diff
@@ -1,26 +1,54 @@
...
+# Common
+common_runs_enabled=true
 common_client_run_name=Common Client
 common_server_run_name=Common Server
 
-# Forge
-forge_version=44.0.1
-//forge_ats_enabled=true
+common_add_to_sources=true
+common_add_to_javadoc=true
```
Comment (jaredlll08) [inline gradle.properties:21]: Once again, why is this even an option

File: gradle.properties [hunk 17]
```diff
@@ -1,26 +1,54 @@
...
+
+# Mappings    Choose from official, parchment, yarn, quilt, quilted_official
+# yarn, quilt, quilted_official: Forge doesn't support yarn and quilt mappings, Forge subproject will be set to Mojang (official) mappings.
+# quilted_official: Use if you want Mojang mappings as your primary mappings, falling back on Quilt mappings for parameters and Javadocs.
+mappings_type=official
```
Comment (jaredlll08) [inline gradle.properties]: This is not going to work and just cause confusion.

The forge project only supports official or parchment, vanillagradle only supports official, so changing this is going to cause the whole project to not compile as soon as any Common code is called.

File: gradle.properties [hunk 18]
```diff
@@ -1,26 +1,54 @@
...
-mod_author=Jared
-mod_id=multiloader
 
 # Gradle
-org.gradle.jvmargs=-Xmx3G
-org.gradle.daemon=false
\ No newline at end of file
+org.gradle.jvmargs=-Xmx4G
+org.gradle.daemon=false
+org.gradle.parallel=true
```
Comment (jaredlll08) [inline gradle.properties]: I have never had a good experience enabling `parallel` with ForgeGradle, fairly sure Lex was very close to banning me after I wasted an hour of both of our time figuring out issues when I forgot I had this enabled.

File: gradle/wrapper/gradle-wrapper.properties [hunk 19]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.1-all.zip
```
Comment (jaredlll08) [inline gradle/wrapper/gradle-wrapper.properties]: Having this be the full `all` distribution is going to cause a lot of people to download a bigger file when they will not use it.

File: settings.gradle [hunk 20]
```diff
@@ -1,20 +1,41 @@
 pluginManagement {
     repositories {
         gradlePluginPortal()
+        mavenCentral()
+        mavenLocal()
```
Comment (jaredlll08) [inline settings.gradle]: Adding mavenLocal() should only be done when you know what you're doing, even the gradle docs for it say it should be avoided.
https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:case-for-maven-local

File: settings.gradle [hunk 21]
```diff
@@ -1,20 +1,41 @@
...
+        maven {
+            name = 'Parchment'
+            url = 'https://maven.parchmentmc.org'
+        }
+    }
+    resolutionStrategy {
+        eachPlugin {
+            if (requested.id.toString() == 'org.spongepowered.mixin') {
+                useModule("org.spongepowered:mixingradle:${requested.version}")
+            }
```
Comment (jaredlll08) [inline settings.gradle]: mixingradle actually has plugin markers now, I need to update the repo to use them, but this is no longer required.

File: Fabric/build.gradle [hunk 22]
```diff
@@ -29,21 +33,98 @@ loom {
...
+        case "quilted_official" :
+            mappings loom.layered {
+                mappings "org.quiltmc:quilt-mappings:$minecraft_version+$quilt_mappings_version:intermediary-v2"
+                officialMojangMappings()
+            }
+            break
+        default :
+            mappings loom.officialMojangMappings()
+            break
+    }
```
Comment (jaredlll08) [inline Fabric/build.gradle]: This whole section of code is never going to work until VanillaGradle supports different mapping types besides mojmap, and I don't believe that will be happening anytime soon.

**Review by jaredlll08 — COMMENTED** [review]
### Preamble
Just a few things before I get into the review:

1) This template is open source, it is meant to be forked and changed to each author's liking, if you don't like how I am doing something in this template, then please fork it, make your changes and build your mod off your fork.

2) My lack of comment on any specific change does not indicate approval of said change, there are a lot of changes here that I am not a fan of, but the only comment I can add to them is "I don't like this", which I don't feel is conductive.

3) I know the template needs some love, it is hard finding time to update it, I'll try and find time after the 1.20 port, or when ForgeGradle works on Gradle 8.

4) I did not look at anything regarding quilt. This template is meant to provide a quick and easy way for you to write code once, and use it on multiple modloaders without having to maintain multiple repositories or branches. So with that in mind, adding quilt seems like it deviates from the purpose of this template, as quilt likes to point out very often, *nearly all* fabric mods run on quilt[^1] and quilt has no plans to change that any time soon[^2], so when I look at it, I do not see a single reason[^3] to use maintain both fabric and quilt, when you can maintain fabric and have it *just work* on quilt.


### Review

There is a lot of changes that are just plain out wrong and will cause issues as soon as you try and load the template, like changing `compileOnly` to `implementation` for adding the `Common` project as a dependency of the `Forge` project, as well as adding support for Fabric to use different type of mappings without those being supported by Forge or Vanilla.

Some changes may have benefit such as adding datagen, however I don't think that really suits this template, not all mods have the need for datagen (I have personally only used it maybe twice in my mods) and not even the Forge MDK or Fabric example mod include it.

I appreciate that you went out of your way to try and improve the template, and I understand that this has been submitted as a draft PR, but with the changes you have made so far, I'm not sure that the end result will be a good fit for this template.


[^1]: You can read more about the mods that don't [here](https://forum.quiltmc.org/t/mod-incompatibility-megathread/261)

[^2]: https://forum.quiltmc.org/t/are-there-any-plans-about-ending-compatibility-for-fabric/284/2

[^3]: Yes I have been told that QSL has a better API than FAPI, but in the context of a multi mod-loader template like this, having a better API doesn't mean the API can be used, unless the QSL APIs have an equivalent API for Forge and Fabric, then using it will lead to your mod having different features on different modloaders, which is not good for users.

**Comment by Nik-Kot** [comment]
Thank you very-very much for the review and for your time.
I am not so experienced im MC modding yet, so some changes are indeed my mistakes, due to not knowing some quirks or core concepts. But some changes you pointed out have good reasoning behind them, let me explain it in matching threads.

I will adress everything you pointed out in the code, fixing my mistakes and implementing your suggestions. My aim is to improve the template.

As for Quilt - I knew that most Fabric mods work with it. The reason for adding support for it is that mod authors might want their mod to behave differenlty specifically on Quilt, be it using its APIs or not. For example due to a bug that is present only on one of the platforms.

File: settings.gradle (hunk 20 above)
Comment (Nik-Kot) [reply settings.gradle:5]: My bad, forgot to remove before committing. Was used to test a locally compiled version of quilt-loom plugin.

File: Common/src/main/resources/pack.mcmeta
Comment (Nik-Kot) [reply Common/src/main/resources/pack.mcmeta:1]: Fabric and Quilt example mods do not feature this file, suggesting that it is not required.

File: Fabric/build.gradle (hunk 6 above)
Comment (Nik-Kot) [reply Fabric/build.gradle:4]: Fabric and Quilt looms conflict when added via root project. Other plugins miss the version specification in subprojects because their versions are defined in root project.

File: Fabric/src/main/resources/examplemod.mixins.json
Comment (Nik-Kot) [reply Fabric/src/main/resources/examplemod.mixins.json:1]: Good suggestion, will do exactly like that.

File: Forge/build.gradle (hunk 10 above)
Comment (Nik-Kot) [reply Forge/build.gradle:112]: Automatically adding Forge run configurations to Intellij once the proect loads. Commented doe to being not a good way to do so.

File: build.gradle (hunk 13 above)
Comment (Nik-Kot) [reply build.gradle:8]: These plugins are incompatible when loaded in the same project (but they can be used separately in different subprojects).

File: build.gradle (hunk 14 above)
Comment (Nik-Kot) [reply build.gradle:20]: Yes they have been. I am going to remove commented out code after final changes, this was left just in case.

File: Forge/build.gradle (hunk 12 above)
Comment (Nik-Kot) [reply Forge/build.gradle:142]: Someone might want only modloader-specific code in corresponding sources/javadoc, and add common sources to a separate .jar.

File: Forge/src/main/java/com/example/examplemod/ExampleMod.java
Comment (Nik-Kot) [reply Forge/src/main/java/com/example/examplemod/ExampleMod.java:1]: Yes, needs to be adressed. Ideally, I would make an API for adding blocks/items/etc to all versions from the Common project.

**Comment by jaredlll08** [comment]
> As for Quilt - I knew that most Fabric mods work with it. The reason for adding support for it is that mod authors might want their mod to behave differenlty specifically on Quilt, be it using its APIs or not. For example due to a bug that is present only on one of the platforms.

In those cases, it should be up to the mod authors to decide what they want to do, if they want to add quilt to the environment, report to quilt / fabric to get the bug fixed or to just not support quilt.

I feel adding quilt when it is not necessarily needed, will lead to extra work for the majority of mod authors supporting both, when in most cases, they should report that bug to quilt, as it's in their best interest to make sure that they don't introduce bugs in fabric mods.

Just note that as I said, I will be cleaning the template up a bit when ForgeGradle 6 releases and we can use Gradle 8, so I don't want you to waste your time now changing things when I will change them again after that happens.

Also we do have a discord server if you want to come and discuss your changes with us there:
https://discord.myceliummod.network

File: Common/src/main/java/com/example/examplemod/CommonInit.java (hunk 3 above)
Comment (Nik-Kot) [reply Common/src/main/java/com/example/examplemod/CommonInit.java:11]: To better represent what the class actually does.

File: gradle.properties (hunk 18 above)
Comment (Nik-Kot) [reply gradle.properties:54]: Will remove.

File: gradle/wrapper/gradle-wrapper.properties (hunk 19 above)
Comment (Nik-Kot) [reply gradle/wrapper/gradle-wrapper.properties:3]: Again, forgot to remove after testing.

File: Fabric/build.gradle (hunk 22 above)
Comment (Nik-Kot) [reply Fabric/build.gradle:91]: This is a draft, I was looking to adding loom to common project, as described in #13

File: Common/build.gradle (hunk 1 above)
Comment (Nik-Kot) [reply Common/build.gradle:26]: Re-added

File: Fabric/src/main/java/com/example/examplemod/mixin/ExampleMixin.java
Comment (Nik-Kot) [reply Fabric/src/main/java/com/example/examplemod/mixin/ExampleMixin.java:1]: To avoid conflict with common mixin which uses this target.

File: Forge/build.gradle (hunk 11 above)
Comment (Nik-Kot) [reply Forge/build.gradle:121]: I will research into this. Was initially done to avoid sourcesets clutter.

**Comment by Nik-Kot** [comment]
> Just note that as I said, I will be cleaning the template up a bit when ForgeGradle 6 releases and we can use Gradle 8, so I don't want you to waste your time now changing things when I will change them again after that happens.

I did not know the project was waiting for ForgeGradle. Really makes sense, concidering latest version of `fabric-loom` now requires Gradle 8.1.
As for upstream changes, I am ready to update my fork according to them.

And I am thinking to making Quilt and option disabled by default. It is way easier for mod devs to delete unneeded files compared to adding QUilt support from scratch and tweaking it so it works. Everything should b in place is case one needs it.

File: Common/build.gradle (hunk 2 above)
Comment (Nik-Kot) [reply Common/build.gradle:32]: Objest's `.clone()` has `protected` acces modifier in Java. Groovy, as an extenson of Java, is supposed to respect its acceess modifiers, but here are mny issues related to that. I decided to go with a `public` constructor just to be safe.

File: gradle.properties (hunk 15 above)
Comment (Nik-Kot) [reply gradle.properties:6]: Will revert back combined notation. Did not know some mods did that.

File: Common/src/main/java/com/example/examplemod/Constants.java (hunk 5 above)
Comment (Nik-Kot) [reply Common/src/main/java/com/example/examplemod/Constants.java:14]: To be more consistent with Forge and Fabric example mods

File: Forge/build.gradle (hunk 9 above)
Comment (Nik-Kot) [reply Forge/build.gradle:47]: Will make this an option with extra logging disabled by default.

File: settings.gradle (hunk 21 above)
Comment (Nik-Kot) [reply settings.gradle:30]: Removed.

File: Common/src/main/java/com/example/examplemod/CommonInit.java (hunk 4 above)
Comment (ghost) [reply Common/src/main/java/com/example/examplemod/CommonInit.java]: Reverted.

File: Common/src/main/resources/pack.mcmeta
Comment (jaredlll08) [reply Common/src/main/resources/pack.mcmeta:1]: This file is required by Minecraft when loading resource/data packs.
https://minecraft.fandom.com/wiki/Pack.mcmeta

File: Forge/build.gradle (hunk 9 above)
Comment (jaredlll08) [reply Forge/build.gradle]: The template does not need extra options, I actually want to cut down on some of the options like if the common runs are enabled or not, and what their name is, it just clutters the build scripts.

What I would suggest for these would be to comment them out, and give a quick, simple single line explanation of why you may want to enable them

File: Forge/build.gradle (hunk 10 above)
Comment (jaredlll08) [reply Forge/build.gradle:120]: This may lead to the runs being inadvertently regenerated when a user doesn't want to generate them.

Also while this template currently doesn't work in eclipse, I do hope that someday it will, in which case you wouldn't want to generate the intelliJ runs

File: Forge/build.gradle (hunk 12 above)
Comment (jaredlll08) [reply Forge/build.gradle:151]: This will cause needless confusion, if the common source is not included in the javadocs and source files, people pulling just the fabric version of the mod from maven, won't have any documentation for common code for example.

There should be no reason for the sources jar to be different than the built jar in terms of content covered.

File: Forge/src/main/java/com/example/examplemod/ExampleMod.java
Comment (jaredlll08) [reply Forge/src/main/java/com/example/examplemod/ExampleMod.java:1]: That will not be accepted into this template, we have a branch that has examples which already features a registration system,  https://github.com/jaredlll08/MultiLoader-Template/tree/1.18.2/examples/MultiLoader-Template-with-registration, if you made your system, it could be added an example there as well.

Not all mods add content to the game, and either modders will need to go through an extra step of removing the code, or they will just leave it in as needless clutter.

File: Fabric/build.gradle (hunk 22 above)
Comment (jaredlll08) [reply Fabric/build.gradle]: We have no plans to change the common project to Loom.

VanillaGradle just works, I have gotten reports that when people try and use loom, it may work for a bit but they always run into issues, but we have had no issues using VanillaGradle.

File: Common/src/main/resources/pack.mcmeta
Comment (Nik-Kot) [reply Common/src/main/resources/pack.mcmeta:1]: Will re-add.

**Comment by jaredlll08** [comment]
> And I am thinking to making Quilt and option disabled by default. It is way easier for mod devs to delete unneeded files compared to adding Quilt support from scratch and tweaking it so it works. Everything should be in place is case one needs it.

To put it plain and simple, quilt is not going to be added to this ***base*** template until quilt has a valid reason to be added, which as detailed in my initial reply, I just don't see a reason for it.

It especially doesn't help that quilt-loom and fabric-loom are incompatible with each other (As you said above).

Saying that, we do already have have an examples branch that features quilt https://github.com/jaredlll08/MultiLoader-Template/tree/1.18.2/examples/MultiLoader-Template-with-quilt

If someone wanted to specifically target quilt, they can use that example as a base, but including it in this base template is going to lead to extra work for the majority of modders who do not have a reason to use quilt.

**Comment by Nik-Kot** [comment]
> To put it plain and simple, quilt is not going to be added to this _**base**_ template until quilt has a valid reason to be added, which as detailed in my initial reply, I just don't see a reason for it.

Ok, I am going to move Quilt to a separate branch once this one is out of draft state.
My plan is support older versions too, using newest compatible build tools and plugins, so I will be looing into 1.18.2 too.

File: gradle.properties [hunk 23]
```diff
@@ -28,6 +28,7 @@ common_add_to_javadoc=true
 forge_version=44.1.23
 forge_ats_enabled=false
 forge_extra_logging=false
+systemProp.net.minecraftforge.gradle.check.gradle=false
```
Comment (jaredlll08) [inline gradle.properties:33]: It is my understanding that Forge offers zero support for running their plugin on unsupported gradle versions, this is not something I want to ship to end users.

File: Common/src/main/java/com/example/examplemod/CommonInit.java [hunk 24]
```diff
@@ -8,7 +8,7 @@
 // import and access the vanilla codebase, libraries used by vanilla, and optionally third party libraries that provide
 // common compatible binaries. This means common code can not directly use loader specific concepts such as Forge events
 // however it will be compatible with all supported mod loaders.
-public class CommonInit {
+public class  CommonInit {
```
Comment (Darkhax) [inline Common/src/main/java/com/example/examplemod/CommonInit.java]: Why are you adding double spacing here, and in other classes?

Comment (Nik-Kot) [reply Common/src/main/java/com/example/examplemod/CommonInit.java:11]: I did not do this manually or intentionally. Probably some Idea formatting rule. I will ind what is causing it and remove.

File: gradle.properties (hunk 23 above)
Comment (Nik-Kot) [reply gradle.properties:33]: This completely is temporary and will be removed once out of draft state (definitely after ForgeGradle 6.0 release). I added this to keep my for up to date and see of any issues arise with Gradle 8

File: Common/src/main/java/com/example/examplemod/CommonInit.java (hunk 24 above)
Comment (Nik-Kot) [reply Common/src/main/java/com/example/examplemod/CommonInit.java]: Fixed

**Comment by jaredlll08** [comment]
I'm going to close this PR as the template has changed a bit since it was made and would most likely require some rewriting of this PR.

Feel free to make a new PR if you update the PR.
